apiserversdk/proxy.go

```diff
@@ -84,3 +92,89 @@ func requireKubeRayService(handler http.Handler, k8sClient *kubernetes.Clientset
 		handler.ServeHTTP(w, r)
 	})
 }
+
+type retryRoundTripper struct {
+	base    http.RoundTripper
+	retries int
+}
+
+func newRetryRoundTripper(base http.RoundTripper, retries int) http.RoundTripper {
+	return &retryRoundTripper{base: base, retries: retries}
+}
+
+func (rrt *retryRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	timeoutCtx, cancel := context.WithTimeout(req.Context(), 30*time.Second)
+	defer cancel()
+
+	req = req.WithContext(timeoutCtx)
+
+	var resp *http.Response
+	var err error
+	for i := 0; i < rrt.retries; i++ {
+		if i > 0 && req.GetBody != nil {
+			var bodyCopy io.ReadCloser
+			bodyCopy, err = req.GetBody()
+			if err != nil {
+				log.Printf("[retry %d/%d] failed to read request body: %v", i+1, rrt.retries, err)
+				return nil, err
+			}
+			req.Body = bodyCopy
+		}
+
+		resp, err = rrt.base.RoundTrip(req)
+		if err == nil {
+			if 200 <= resp.StatusCode && resp.StatusCode < 300 {
+				// Successful status code
+				return resp, nil
+			} else if shouldRetry(resp.StatusCode) {
+				// Error status code that should retry
+				log.Printf("[retry %d/%d] request failed with retryable error status: %s", i+1, rrt.retries, resp.Status)
+			} else {
+				// Error status code that should not retry
+				log.Printf("[retry %d/%d] request failed with received non-retryable error status: %s", i+1, rrt.retries, resp.Status)
+				return resp, nil
+			}
+		} else {
```

Comment:
let's do the else branch first (if-else branch for whether taking err for `GetBody`)